dffml/util/testing/source.py

```diff
@@ -60,40 +59,78 @@ class SourceTest(abc.ABC):
+                        repo.data.prediction.classification, "feedface"
+                    )
+                    self.assertEqual(repo.data.prediction.confidence, 0.42)
+                with self.subTest(src_url=empty_src_url):
+                    repo = await sourceContext.repo(empty_src_url)
+                    self.assertFalse(repo.data.prediction.classification)
+                    self.assertFalse(repo.data.prediction.confidence)
+                with self.subTest(both=[full_src_url, empty_src_url]):
+                    repos = {
+                        repo.src_url: repo
+                        async for repo in sourceContext.repos()
+                    }
+                    self.assertIn(full_src_url, repos)
+                    self.assertIn(empty_src_url, repos)
+                    self.assertEqual(
+                        repos[full_src_url].features(), full_repo.features()
+                    )
+                    self.assertEqual(
+                        repos[empty_src_url].features(), empty_repo.features()
+                    )
+
+
+class FileSourceTest(SourceTest):
+    """
+    Test case class used to test a Source implementation. Subclass from and set
+    the SOURCE property to run tests on that source.
+
+    >>> from dffml.source.file import FileSourceConfig
+    >>> from dffml.source.json import JSONSource
+    >>> from dffml.util.testing.source import FileSourceTest
+    >>> from dffml.util.asynctestcase import AsyncTestCase
+    >>> class TestJSONSource(FileSourceTest, AsyncTestCase):
+    >>>     async def setUpSource(self):
+    >>>         return JSONSource(FileSourceConfig(filename=self.testfile))
+    """
+
+    async def test_update(self):
+        with tempfile.TemporaryDirectory() as testdir:
+            with self.subTest(extension=None):
+                self.testfile = os.path.join(testdir, str(random.random))
+                await super().test_update()
+            for extension in ["xz", "gz", "bz2", "lzma", "zip"]:
+                with self.subTest(extension=extension):
+                    self.testfile = os.path.join(
+                        testdir, str(random.random) + "." + extension
+                    )
+                    await super().test_update()
+
+    async def test_label(self):
+        with tempfile.TemporaryDirectory() as testdir:
+            self.testfile = os.path.join(testdir, str(random.random))
+            unlabeled = await self.setUpSource()
+            labeled = await self.setUpSource()
+            labeled.config = labeled.config._replace(label="somelabel")
+            async with unlabeled:
+                async with unlabeled() as uctx:
+                    await uctx.update(
+                        Repo("0", data={"features": {"life": 42}})
+                    )
+            async with labeled:
+                async with labeled() as lctx:
+                    repo = await lctx.repo("0")
+                    self.assertNotIn("life", repo.features())
```

Comment:
Add something which opens both at the same time then closes, then opens again and verifiy's they are still there.

```python
            async with unlabeled, labeled:
                async with unlabeled() as uctx, labeled() as lctx:
                    await uctx.update(
                        Repo("0", data={"features": {"feed": 1}})
                    )
                    await lctx.update(
                        Repo("0", data={"features": {"face": 2}})
                    )
            async with unlabeled, labeled:
                async with unlabeled() as uctx, labeled() as lctx:
                    repo = await uctx.repo("0")
                    self.assertNotIn("feed", repo.features())
                    repo = await lctx.repo("0")
                    self.assertNotIn("face", repo.features())
```

do this first, it shouldn't pass, change the code, it should pass